### DISCUSSION TITLE
MDX v3 - Upgrade Support

### DISCUSSION THREAD

**Discussion by slorber** [facebook/docusaurus#9053] [General]
Docusaurus v3 will use MDX v3.

This is a breaking change that might require you to adjust your docs. 

This GitHub discussion is a place to ask for help in case our migration notes are not enough, and you can't find a solution. 

## Other useful support links

- [Docusaurus v3 - Upgrade Support](https://github.com/facebook/docusaurus/discussions/9336)
- [Docusaurus v3 - Discord channel #migration-v2-to-v3](https://discord.com/channels/398180168688074762/1154771869094912090)
- [MDX v3 - Upgrade Support](https://github.com/facebook/docusaurus/discussions/9053)
- [MDX v3 - Remark/Rehype Plugins Support](https://github.com/facebook/docusaurus/discussions/9337)
- [MDX v3 - Discord channel #migration-mdx-v3](https://discord.com/channels/398180168688074762/1116724556976111616)

## Before asking for support

Before asking for support, make sure you:
- have read carefully the [Docusaurus MDX v2 PR](https://github.com/facebook/docusaurus/pull/8288) instructions and understood the various compatibility options we provide
- have read carefully the official [MDX v2](https://mdxjs.com/blog/v2/) and [MDX v3](https://mdxjs.com/blog/v3/) release blog posts
- have read carefully the [MDX v2](https://mdxjs.com/migrating/v2/) and [MDX v3](https://mdxjs.com/migrating/v3/) migration guides
- have tested your Markdown code in the [MDX playground](https://mdxjs.com/playground/): this might help you figure out the problem on your own

To ask for support, please:
- Check if someone else already asked a similar question
- Create a minimal repro using [docusaurus.new](https://docusaurus.new)
- Give us the fs path of the problematic Markdown file, and the URL pathname to access it
- Spend some time explaining your problem clearly in a structured and concise way
- Paste full, clear error messages and provide screenshots if appropriate

It's not scalable to provide done-for-you support for all Docusaurus users looking to upgrade. Any support request not respecting those constraints will possibly be ignored. 

**If you want help, you need to invest your own time so that helping you is easy.**

Alternatively, you can look for a paid [Docusaurus Service Provider](https://github.com/facebook/docusaurus/discussions/9281) to help you.

**Comment by ntucker** [comment]
Is there a script to easily rename all our .md files to .mdx since this is required in the upgrade? This will also need to rename all links since docusaurus has recommended we use filenames for links

**Reply by ntucker** [reply]
Or perhaps an option to disable commonmark completely?

**Reply by slorber** [reply]
Renaming all files from `.md` to `.mdx` is not required for the upgrade. 

But it is recommended to use the `.mdx` extension when you use MDX / JSX. 
Not by Docusaurus, but by the MDX creator.
Using the correct extension allows related tools to work better (IDE, Prettier, ESLint, syntax highlighting, GitHub preview...)

As part of https://github.com/facebook/docusaurus/pull/9097, there is now an option to customize how Docusaurus process md/mdx files by default. For better compatibility with Docusaurus v2, Docusaurus v3 will parse everything with MDX by default, but you can opt-in for CommonMark parsing.

You can opt-in for CommonMark parsing by using `siteConfig.markdown.format: 'md' | 'mdx' | 'detect'`, and on a per-file basis using `format: 'md' | 'mdx' | 'detect'`.

---

There is no automated script to migrate extensions to the `.mdx`, but that could be nice if someone in the community worked on one.

I think I have used that locally for the Docusaurus site itself. No guarantee it will work for you, but you can use it as an inspiration

CLI to rename files: 

```bash
find website/docs -type f -name "*.md" -exec rename 's/\.md$/.mdx/' '{}' \;`
```

IDE/Intellij replaceAll regexp to replace markdown link paths:

```
\(https[^ ]*?\.mdx(#[^ ]+)?\)


From: \.md(?<hash>#(.*?))?\)
To: .mdx${hash})


From: \((?<url>(?!http)[^ ]*?)(?<ext>\.md)(?<hash>#[^ ]+)?\)
To: (${url}.mdx${hash})


Find .md refs in docs: \.md[^ x)#]

```

Again these are rough notes that I took while doing the migration and are just an inspiration. If someone can come up with proper instructions that work for all OS/IDEs that can be nice. I will try to see if I can write a process myself.

**Comment by ntucker** [comment]
Anyone get

```
ERROR in ../docs/core/concepts/normalization.md
Module build failed (from ./node_modules/@docusaurus/mdx-loader/lib/index.js):
Error: MDX compilation failed for file "/home/ntucker/src/rest-hooks/docs/core/concepts/normalization.md"
Cause: chunks[startIndex].slice is not a function
Details:
{}
    at Object.mdxLoader (/home/ntucker/src/rest-hooks/website/node_modules/@docusaurus/mdx-loader/lib/loader.js:111:25)
```

<details><summary>normalization.md</summary>

---
title: Entity and Data Normalization
sidebar_label: Data Normalization
---

<head>
  <meta name="docsearch:pagerank" content="40"/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import LanguageTabs from '@site/src/components/LanguageTabs';

[Entities](/rest/api/Entity) have a primary key. This enables easy access via a lookup table.
This makes it easy to find, update, create, or delete the same data - no matter what
endpoint it was used in.



<Tabs
defaultValue="State"
values={[
{ label: 'State', value: 'State' },
{ label: 'Response', value: 'Response' },
{ label: 'Endpoint', value: 'Endpoint' },
{ label: 'Entity', value: 'Entity' },
{ label: 'React', value: 'React' },
]}>
<TabItem value="State">

![Entities cache](/img/entities.png)

</TabItem>
<TabItem value="Response">

```json
[
  { "id": 1, "title": "this is an entity" },
  { "id": 2, "title": "this is the second entity" }
]
```

</TabItem>
<TabItem value="Endpoint">

```typescript
const PresentationList = new Endpoint(
  () => fetch(`/presentations`).then(res => res.json()),
  { schema: [PresentationEntity] },
);
```

</TabItem>
<TabItem value="Entity">

```typescript
class PresentationEntity extends Entity {
  readonly id: string = '';
  readonly title: string = '';

  pk() {
    return this.id;
  }
}
```

</TabItem>
<TabItem value="React">

```tsx
export function PresentationsPage() {
  const presentation = useSuspense(PresentationList, {});
  return presentation.map(presentation => (
    <div key={presentation.pk()}>{presentation.title}</div>
  ));
}
```

</TabItem>
</Tabs>

Extracting entities from a response is known as `normalization`. Accessing a response reverses
the process via `denormalization`.

:::info Global Referential Equality

Using entities expands Rest Hooks' global referential equality guarantee beyond the granularity of
an entire endpoint response.

:::

## Mutations and Dynamic Data

When an endpoint changes data, this is known as a [side effect](/rest/guides/rpc). Marking an endpoint with [sideEffect: true](/rest/api/Endpoint#sideeffect)
tells Rest Hooks that this endpoint is not idempotent, and thus should not be allowed in hooks
that may call the endpoint an arbitrary number of times like [useSuspense()](../api/useSuspense.md) or [useFetch()](../api/useFetch.md)

By including the changed data in the endpoint's response, Rest Hooks is able to able to update
any entities it extracts by specifying the schema.

<Tabs
defaultValue="Create"
values={[
{ label: 'Create', value: 'Create' },
{ label: 'Update', value: 'Update' },
{ label: 'Delete', value: 'Delete' },
]}>
<TabItem value="Create">

```typescript
import { RestEndpoint } from '@rest-hooks/rest';

const todoCreate = new RestEndpoint({
  urlPrefix: 'https://jsonplaceholder.typicode.com',
  path: '/todos',
  method: 'POST',
  schema: Todo,
});
```

<details><summary><b>Example Usage</b></summary>

```tsx
import { useController } from '@rest-hooks/react';

export default function NewTodoForm() {
  const ctrl = useController();
  return (
    <Form onSubmit={e => ctrl.fetch(todoCreate, new FormData(e.target))}>
      <FormField name="title" />
    </Form>
  );
}
```

</details>

</TabItem>
<TabItem value="Update">

```typescript
import { RestEndpoint } from '@rest-hooks/rest';

const todoUpdate = new RestEndpoint({
  urlPrefix: 'https://jsonplaceholder.typicode.com',
  path: '/todos/:id',
  method: 'PUT',
  schema: Todo,
});
```

<details><summary><b>Example Usage</b></summary>

```tsx
import { useController } from '@rest-hooks/react';

export default function UpdateTodoForm({ id }: { id: number }) {
  const todo = useSuspense(todoDetail, { id });
  const ctrl = useController();
  return (
    <Form
      onSubmit={e => ctrl.fetch(todoUpdate, { id }, new FormData(e.target))}
      initialValues={todo}
    >
      <FormField name="title" />
    </Form>
  );
}
```

</details>

</TabItem>
<TabItem value="Delete">

```typescript
import { schema, RestEndpoint } from '@rest-hooks/rest';

const todoDelete = new RestEndpoint({
  urlPrefix: 'https://jsonplaceholder.typicode.com',
  path: '/todos/:id',
  method: 'DELETE',
  schema: new schema.Delete(Todo),
});
```

<details collapsed><summary><b>Example Usage</b></summary>

```tsx
import { useController } from '@rest-hooks/react';

export default function TodoWithDelete({ todo }: { todo: Todo }) {
  const ctrl = useController();
  return (
    <div>
      {todo.title}
      <button onClick={() => ctrl.fetch(todoDelete, { id: todo.id })}>
        Delete
      </button>
    </div>
  );
}
```

</details>

</TabItem>
</Tabs>

:::info

Mutations automatically update the normalized cache, resulting in consistent and fresh data.

:::

## Schema

Schemas are a declarative definition of how to [process responses](/rest/api/schema)

- [where](/rest/api/schema) to expect [Entities](/rest/api/Entity)
- Classes to [deserialize fields](/rest/guides/network-transform#deserializing-fields)

```typescript
import { RestEndpoint } from '@rest-hooks/rest';

const getTodoList = new RestEndpoint({
  urlPrefix: 'https://jsonplaceholder.typicode.com',
  path: '/todos',
  // highlight-next-line
  schema: [Todo],
});
```

Placing our [Entity](/rest/api/Entity) `Todo` in an array, tells Rest Hooks to expect
an array of `Todos`.

Aside from array, there are a few more 'schemas' provided for various patterns. The first two (Object and Array)
have shorthands of using object and array literals.

- [Object](/rest/api/Object): maps with known keys
- [Array](/rest/api/Array): variably sized arrays
- [Union](/rest/api/Union): select from many different types
- [Values](/rest/api/Values): maps with any keys - variably sized
- [Invalidate](/rest/api/Invalidate): remove an entity

[Learn more](/rest/api/schema)

### Nesting

Additionally, [Entities](/rest/api/Entity) themselves can specify [nested schemas](/rest/guides/relational-data)
by specifying a [static schema](/rest/api/Entity#schema) member.

<Tabs
defaultValue="Entity"
values={[
{ label: 'Entity', value: 'Entity' },
{ label: 'Response', value: 'Response' },
]}>
<TabItem value="Entity">

```typescript
import { Entity } from '@rest-hooks/endpoint';

class Todo extends Entity {
  readonly id: number = 0;
  readonly user: User = User.fromJS({});
  readonly title: string = '';
  readonly completed: boolean = false;

  pk() {
    return `${this.id}`;
  }

  // highlight-start
  static schema = {
    user: User,
  };
  // highlight-end
}

class User extends Entity {
  readonly id: number = 0;
  readonly username: string = '';

  pk() {
    return `${this.id}`;
  }
}
```

</TabItem>
<TabItem value="Response">

```json
{
  "id": 5,
  "user": {
    "id": 10,
    "username": "bob"
  },
  "title": "Write some Entities",
  "completed": false
}
```

</TabItem>
</Tabs>

[Learn more](/rest/guides/relational-data)

### Data Representations

Additionally, any `newable` class that has a toJSON() method, can be [used as a schema](/rest/guides/network-transform#deserializing-fields). This will simply construct the object during denormalization.
This might be useful with representations like [bignumber](https://mikemcl.github.io/bignumber.js/)

```ts
import { Entity } from '@rest-hooks/endpoint';

class Todo extends Entity {
  readonly id: number = 0;
  readonly user: User = User.fromJS({});
  readonly title: string = '';
  readonly completed: boolean = false;
  // highlight-next-line
  readonly dueDate: Date = new Date(0);

  pk() {
    return `${this.id}`;
  }

  static schema = {
    user: User,
    // highlight-next-line
    dueDate: Date,
  };
}
```

:::info

Due to the global referential equality guarantee - construction of members only occurs once
per update.

:::

</details>

with the canary build?

Is this due to it using commonmark?

**Reply by slorber** [reply]
This is a bug with multi-line HTML comments.

Reported here: https://github.com/facebook/docusaurus/issues/9084

Will be fixed soon, and there's a temporary workaround
Edit: now fixed in latest canary

**Comment by johnnyreilly** [comment]
Just linking in this issue here: https://github.com/facebook/docusaurus/issues/9081#issuecomment-1601350219

I'd love to see a guide on how to migrate remark / rehype plugins to Docusaurus v3 / MDX v2

**Reply by slorber** [reply]
It's not super easy to write a generic guide for that. 

You have to read the MDX v2 docs and understand how the whole Unified ecosystem works.

I created a dedicated issue to help Remark/Rehype plugin authors, with some useful notes to put you on track.

[MDX v2 - Remark/Rehype Plugins Support](https://github.com/facebook/docusaurus/discussions/9337)

**Comment by ntucker** [comment]
```bash
Error: MDX compilation failed for file "/home/ntucker/src/rest-hooks/docs/graphql/api/validateRequired.md"
Cause: Expected a closing tag for `</summary>` (57:1-57:20) before the end of `paragraph`
Details:
{
  "name": "57:1-57:55",
  "message": "Expected a closing tag for `</summary>` (57:1-57:20) before the end of `paragraph`",
  "reason": "Expected a closing tag for `</summary>` (57:1-57:20) before the end of `paragraph`",
  "line": 57,
  "column": 1,
  "position": {
    "start": {
      "line": 57,
      "column": 1,
      "offset": 1389,
      "_index": 0,
      "_bufferIndex": 0
    },
    "end": {
      "line": 57,
      "column": 55,
      "offset": 1443,
      "_index": 1,
      "_bufferIndex": -1
    }
  },
  "source": "mdast-util-mdx-jsx",
  "ruleId": "end-tag-mismatch"
}
```

However the only 'summary' shows up as `<details collapsed><summary><b>exclude()</b></summary>`. There is no paragraph and it is correctly closed...

Also this is on line 61...so the line numbers are off?

Site: https://github.com/data-client/rest-hooks in /website (using 'canary' to replace existing version of docusaurus)

Full file below

<details><summary><b>validateRequired.md</b></summary>

---
title: validateRequired
---

```ts
function validateRequired(processedEntity: any, requiredDefaults: Record<string, unknown>): string | undefined;
```

Returns a string message if any keys of `requiredDefaults` are missing in `processedEntity`. This
can be used to [validate](./Entity.md#validate) fields that must be provided.

```ts
class CustomBaseEntity extends Entity {
  static validate(processedEntity) {
    return validateRequired(processedEntity, this.defaults) || super.validate(processedEntity);
  }
}
```

## Partial/full results

This can be useful to automatically validate for [partial results](/docs/concepts/validation#partial-results)

```ts
class SummaryAnalysis extends Entity {
  readonly id: string = '';
  readonly createdAt: Date = new Date(0);
  readonly meanValue: number = 0;
  readonly title: string = '';

  pk() {
    return this.id;
  }
}

class FullAnalysis extends SummaryAnalysis {
  readonly graph: number[] = [];
}
```

## Optional fields

In case we have a field that won't always be present (like `lastRun` here), we can simply
'exclude' it from the fields we require.

```ts
class FullAnalysis extends SummaryAnalysis {
  readonly graph: number[] = [];
  readonly lastRun?: Date = new Date(0);

  static schema = {
    lastRun: Date,
  }

  static validate(processedEntity) {
    return validateRequired(processedEntity, exclude(this.defaults, ['lastRun']));
  }
}
```

<details collapsed><summary><b>exclude()</b></summary>

```ts title="exclude"
function exclude<O extends Record<string, unknown>>(
  obj: O,
  keys: string[],
): Partial<O> {
  const r: any = {};
  Object.keys(obj).forEach(k => {
    if (!keys.includes(k)) r[k] = obj[k];
  });
  return r;
}
```

</details>

### Full results only have optional fields

In case every field of the 'full' resource was optional:

```ts
class FullAnalysis extends SummaryAnalysis {
  readonly graph?: number[] = [];
  readonly lastRun?: Date = new Date(0);

  static schema = {
    lastRun: Date,
  }

  static validate(processedEntity) {
    return validateRequired(processedEntity, exclude(this.defaults, ['graph', 'lastRun']));
  }
}
```

This code would not successfully know to fetch the 'full' resource if the summary is already provided.
There would be no way of knowing whether the fields simply don't exist for that data, or were not fetched.
In this case, it is best to provide a `null` default for *at least* one field.

```ts
class FullAnalysis extends SummaryAnalysis {
  readonly graph: number[] = null;
  readonly lastRun?: Date = new Date(0);

  static schema = {
    lastRun: Date,
  }

  static validate(processedEntity) {
    return validateRequired(processedEntity, exclude(this.defaults, ['lastRun']));
  }
}
```

This enables the client to understand whether the 'full' resource has been fetched at all.

</details>

**Reply by ntucker** [reply]
I reduced the entire file to just the details element and it still complained:

```mdx
<details collapsed><summary><b>exclude()</b></summary>

hi

</details>
```

**Reply by ntucker** [reply]
Figured this out on my own, but I still cannot find this in the [migration docs](https://mdxjs.com/migrating/v2)....

```mdx
<details collapsed>
<summary><b>exclude()</b></summary>

hi

</details>
```

Needs a newline now for some reason

**Reply by slorber** [reply]
@ntucker if you can reproduce this in the [MDX Playground](https://mdxjs.com/playground/) (it's the case here) and don't understand the MDX behavior, then the best is to search MDX issues and eventually open a question on that repository directly: if MDX has a certain behavior, Docusaurus and other tools using it will usually have it too.

In this case, I have no idea why it doesn't compile 😅 The minimal case seems:

```mdx
<details><summary>summary</summary>

</details>
```

Any idea @wooorm ?

<img width="977" alt="CleanShot 2023-06-29 at 16 35 24@2x" src="https://github.com/facebook/docusaurus/assets/749374/a2bac206-ff50-4ce5-b45d-578ce2c1ca4f">

**Reply by wooorm** [reply]
It’s unclear. The details element here wants to have blocks (flow content). The summary element wants to have inline (text content). It can’t combine like that. 

Put text and tags on the same line, or not, don’t mix things:

```mdx
<x><y>z</y></x>
```

```mdx
<x>
  <y>z</y>
</x>
```

```mdx
<x>
  <y>
    z
  </y>
</x>
```

**Comment by ntucker** [comment]
Is a code block metastring not parsed anymore? For

\```typescript title="api/Feed.ts"
const a = 1;
\```

I get props of:

```js
{
className:"language-typescript"
metastring: "title=\"api/Feed.ts\""
}
```

**Reply by slorber** [reply]
The props you get are expected and afaik didn't change.

> Is a code block metastring not parsed anymore? For

In Docusaurus v3.0 the metastring remains parsed inside the CodeBlock theme component, like it used to be in v2.

Your code snippet should display as before with a title despite not having a `title` prop:

![CleanShot 2023-06-29 at 16 45 14@2x](https://github.com/facebook/docusaurus/assets/749374/a0428e15-8cd6-4ad4-9244-faed1cd1ea82)

I don't understand what lead you to think anything changed recently.

**Reply by ntucker** [reply]
This is what I get when inspecting a component created by block comments in the current version:

```js
{
  children:
    "//the javascript code is here",
  className: 'language-tsx',
  mdxType: 'code',
  metastring: 'title="api/Feed.ts"',
  originalType: 'code',
  parentName: 'pre',
  title: 'api/Feed.ts',
}
```

I use this to enhance the block comments turning them into a live preview rather than just the code itself. See the live previews at https://dataclient.io/docs/api/useSuspense as example [mdx source](https://github.com/data-client/rest-hooks/blob/master/docs/core/api/useSuspense.md?plain=1#L35)

Obviously nothing is stopping me from figuring out how to parse it myself, etc. However, I am wondering if that code was moved somewhere and if this is intentional? This will also be obvious when people inspect things using the react dev tools in the browser.

**Comment by MilenkoPMarkovic** [comment]
I have problem
```
Error while compiling file ../docs/concepts/data_versioning.mdx (Line=157 Column=2)
Details: Could not parse expression with acorn
```

line
`{@codeBlock: TestEpochManager.defineCustomEpochPolicy}`

How to change the line?
